File: frontend/src/api/chains.ts
import type { CommonListQuery, PageResponse } from "@lynx/local-console-shared";

import { buildQueryString, fetchJson } from "./client";

export interface ChainSummary {
  chainId: string;
  sessionKey: string;
  channelProfile?: string;
  channelId?: string;
  conversationId?: string;
  status?: string;
  recentIdentity: string[];
  recentSensitive: string[];
  recentDenials: string[];
  recentApprovals: string[];
  recentTools: string[];
  recentTaintReads: string[];
  recentEvasions: string[];
  activeGrantId: string;
  pendingApproval: string;
  coveredPrompts: ChainCoveredPrompt[];
  promptCount: number;
}

export interface ChainCoveredPrompt {
  qaRecordId: string;
  runId?: string;
  userPromptExcerpt: string;
  riskLevel?: string;
  startedAtMs?: number;
  status?: string;
}

export interface ChainListQuery extends CommonListQuery {
  channelProfile?: string;
  conversationId?: string;
}

type ChainListPayload = ChainSummary[] | Partial<PageResponse<ChainSummary>>;

function normalizeStringArray(value: unknown): string[] {
  return Array.isArray(value)
    ? value.filter((item): item is string => typeof item === "string")
    : [];
}

function normalizeCoveredPrompts(value: unknown): ChainCoveredPrompt[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value
    .filter(
      (item): item is Partial<ChainCoveredPrompt> =>
        Boolean(item) && typeof item === "object",
    )
    .map((item) => ({
      qaRecordId: typeof item.qaRecordId === "string" ? item.qaRecordId : "",
      runId: typeof item.runId === "string" ? item.runId : undefined,
      userPromptExcerpt:
        typeof item.userPromptExcerpt === "string"
          ? item.userPromptExcerpt
          : "",
      riskLevel:
        typeof item.riskLevel === "string" ? item.riskLevel : undefined,
      startedAtMs:
        typeof item.startedAtMs === "number" ? item.startedAtMs : undefined,
      status: typeof item.status === "string" ? item.status : undefined,
    }))
    .filter((item) => item.qaRecordId || item.userPromptExcerpt);
}

function normalizeChainSummary(item: ChainSummary): ChainSummary {
  const coveredPrompts = normalizeCoveredPrompts(item.coveredPrompts);
  return {
    ...item,
    channelProfile: typeof item.channelProfile === "string" ? item.channelProfile : undefined,
    channelId: typeof item.channelId === "string" ? item.channelId : undefined,
    conversationId: typeof item.conversationId === "string" ? item.conversationId : undefined,
    status: typeof item.status === "string" ? item.status : undefined,
    recentIdentity: normalizeStringArray(item.recentIdentity),
    recentSensitive: normalizeStringArray(item.recentSensitive),
    recentDenials: normalizeStringArray(item.recentDenials),
    recentApprovals: normalizeStringArray(item.recentApprovals),
    recentTools: normalizeStringArray(item.recentTools),
    recentTaintReads: normalizeStringArray(item.recentTaintReads),
    recentEvasions: normalizeStringArray(item.recentEvasions),
    coveredPrompts,
    promptCount:
      typeof item.promptCount === "number"
        ? item.promptCount
        : coveredPrompts.length,
  };
}

function normalizeChainList(
  payload: ChainListPayload,
  query: ChainListQuery,
): PageResponse<ChainSummary> {
  const rawItems = Array.isArray(payload) ? payload : payload.items ?? [];
  const items = rawItems.map(normalizeChainSummary);
  if (Array.isArray(payload)) {
    const pageSize = query.pageSize ?? payload.length;
    return {
      items,
      pageNum: query.pageNum ?? 1,
      pageSize,
      total: payload.length,
      totalPages: payload.length === 0 ? 0 : Math.ceil(payload.length / pageSize),
    };
  }

  const pageSize = payload.pageSize ?? query.pageSize ?? items.length;
  const total = payload.total ?? items.length;
  return {
    items,
    pageNum: payload.pageNum ?? query.pageNum ?? 1,
    pageSize,
    total,
    totalPages: payload.totalPages ?? (total === 0 ? 0 : Math.ceil(total / pageSize)),
  };
}

export async function listChains(
  query: ChainListQuery = {},
): Promise<PageResponse<ChainSummary>> {
  const payload = await fetchJson<ChainListPayload>(
    `/chains${buildQueryString(query)}`,
  );
  return normalizeChainList(payload, query);
}

export function getChain(chainId: string): Promise<ChainSummary> {
  return fetchJson<ChainSummary>(`/chains/${encodeURIComponent(chainId)}`).then(
    normalizeChainSummary,
  );
}

File: backend/internal/repo/chains.go
package repo

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strings"

	"github.com/openclaw/lynx-guardian/backend/internal/api"
	"github.com/openclaw/lynx-guardian/backend/internal/service"
)

type ChainRepository struct {
	db *sql.DB
}

type ChainListQuery struct {
	Q              *string
	ChannelProfile *string
	ConversationID *string
	SessionKey     *string
	RequesterID    *string
	PageNum        *int
	PageSize       *int
	Limit          *int
}

func NewChainRepository(db *sql.DB) *ChainRepository {
	return &ChainRepository{db: db}
}

func (r *ChainRepository) Upsert(ctx context.Context, input api.ChainUpdateRequest, summary api.ChainSummary, now string) error {
	normalizeChainSummary(&summary)
	summaryJSON := jsonText(summary, "{}")
	_, err := r.db.ExecContext(ctx, `
		INSERT INTO chains (
			id, chain_id, session_key, channel_profile, channel_id, conversation_id,
			requester_id, requester_ou_id, status, summary_json, active_grant_id,
			pending_approval_id, created_at, updated_at, ended_at
		)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
		ON CONFLICT(chain_id) DO UPDATE SET
			session_key = excluded.session_key,
			channel_profile = excluded.channel_profile,
			channel_id = excluded.channel_id,
			conversation_id = excluded.conversation_id,
			requester_id = excluded.requester_id,
			requester_ou_id = excluded.requester_ou_id,
			status = excluded.status,
			summary_json = excluded.summary_json,
			active_grant_id = excluded.active_grant_id,
			pending_approval_id = excluded.pending_approval_id,
			updated_at = excluded.updated_at,
			ended_at = excluded.ended_at`,
		input.ChainID,
		input.ChainID,
		input.SessionKey,
		input.ChannelProfile,
		input.ChannelID,
		input.ConversationID,
		input.RequesterID,
		input.RequesterOuID,
		chainStatus(input.EventType),
		summaryJSON,
		summary.ActiveGrantID,
		summary.PendingApproval,
		now,
		now,
		endedAt(input.EventType, now),
	)
	return err
}

func (r *ChainRepository) AppendEvent(ctx context.Context, input api.ChainUpdateRequest, now string) error {
	_, err := r.db.ExecContext(ctx, `
		INSERT INTO chain_events (
			id, chain_id, event_type, hook, risk_level, action, tool_name,
			target_uri, content_excerpt, metadata_json, created_at
		)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		fmt.Sprintf("%s-%s-%s", input.ChainID, input.EventType, now),
		input.ChainID,
		input.EventType,
		input.Hook,
		input.RiskLevel,
		input.Action,
		input.ToolName,
		input.TargetURI,
		truncate(input.Content, 240),
		jsonText(input.Metadata, "{}"),
		now,
	)
	return err
}

func (r *ChainRepository) Get(ctx context.Context, chainID string) (api.ChainSummary, error) {
	var summaryJSON string
	var channelProfile string
	var channelID string
	var conversationID string
	err := r.db.QueryRowContext(ctx, `
		SELECT
			summary_json,
			COALESCE(channel_profile, ''),
			COALESCE(channel_id, ''),
			COALESCE(conversation_id, '')
		FROM chains
		WHERE chain_id = ?`,
		chainID,
	).Scan(&summaryJSON, &channelProfile, &channelID, &conversationID)
	if errors.Is(err, sql.ErrNoRows) {
		return api.ChainSummary{}, nil
	}
	if err != nil {
		return api.ChainSummary{}, err
	}
	var summary api.ChainSummary
	unmarshalJSONText(summaryJSON, &summary)
	applyChainRouteMetadata(&summary, channelProfile, channelID, conversationID)
	normalizeChainSummary(&summary)
	if err := r.loadPromptCoverage(ctx, &summary); err != nil {
		return api.ChainSummary{}, err
	}
	applyChainGroupingRelation(&summary)
	return summary, nil
}

func (r *ChainRepository) List(ctx context.Context, query ChainListQuery) (service.PageResponse[api.ChainSummary], error) {
	page := service.ResolvePageRequest(query.PageNum, query.PageSize, query.Limit)
	filter := &Filter{}
	filter.AppendTextSearch([]string{
		"chain_id",
		"session_key",
		"channel_profile",
		"channel_id",
		"conversation_id",
		"requester_id",
		"requester_ou_id",
		"status",
		"summary_json",
		"active_grant_id",
		"pending_approval_id",
	}, query.Q)
	filter.AppendEquals("channel_profile", query.ChannelProfile)
	filter.AppendEquals("conversation_id", query.ConversationID)
	filter.AppendEquals("session_key", query.SessionKey)
	filter.AppendEquals("requester_id", query.RequesterID)

	total, err := countRowsContext(ctx, r.db, "chains", filter)
	if err != nil {
		return service.PageResponse[api.ChainSummary]{}, err
	}

	rows, err := r.db.QueryContext(ctx, `
		SELECT
			summary_json,
			COALESCE(channel_profile, ''),
			COALESCE(channel_id, ''),
			COALESCE(conversation_id, '')
		FROM chains `+filter.Where()+`
		ORDER BY updated_at DESC, chain_id DESC
		LIMIT ? OFFSET ?`,
		append(filter.Params(), page.PageSize, page.Offset)...,
	)
	if err != nil {
		return service.PageResponse[api.ChainSummary]{}, err
	}
	defer rows.Close()

	out := make([]api.ChainSummary, 0)
	for rows.Next() {
		var summaryJSON string
		var channelProfile string
		var channelID string
		var conversationID string
		if err := rows.Scan(&summaryJSON, &channelProfile, &channelID, &conversationID); err != nil {
			return service.PageResponse[api.ChainSummary]{}, err
		}
		var summary api.ChainSummary
		unmarshalJSONText(summaryJSON, &summary)
		applyChainRouteMetadata(&summary, channelProfile, channelID, conversationID)
		normalizeChainSummary(&summary)
		out = append(out, summary)
	}
	if err := rows.Err(); err != nil {
		return service.PageResponse[api.ChainSummary]{}, err
	}
	for index := range out {
		if err := r.loadPromptCoverage(ctx, &out[index]); err != nil {
			return service.PageResponse[api.ChainSummary]{}, err
		}
		applyChainGroupingRelation(&out[index])
	}
	return service.BuildPageResponse(out, total, page), nil
}

func applyChainRouteMetadata(summary *api.ChainSummary, channelProfile string, channelID string, conversationID string) {
	summary.ChannelProfile = channelProfile
	summary.ChannelID = channelID
	summary.ConversationID = conversationID
}

func normalizeChainSummary(summary *api.ChainSummary) {
	if summary.RecentIdentity == nil {
		summary.RecentIdentity = []string{}
	}
	if summary.RecentSensitive == nil {
		summary.RecentSensitive = []string{}
	}
	if summary.RecentDenials == nil {
		summary.RecentDenials = []string{}
	}
	if summary.RecentApprovals == nil {
		summary.RecentApprovals = []string{}
	}
	if summary.RecentTools == nil {
		summary.RecentTools = []string{}
	}
	if summary.RecentTaintReads == nil {
		summary.RecentTaintReads = []string{}
	}
	if summary.RecentEvasions == nil {
		summary.RecentEvasions = []string{}
	}
	if summary.CoveredPrompts == nil {
		summary.CoveredPrompts = []api.ChainCoveredPrompt{}
	}
	summary.PromptCount = len(summary.CoveredPrompts)
}

func applyChainGroupingRelation(summary *api.ChainSummary) {
	parts := make([]string, 0, 5)
	if summary.SessionKey != "" {
		parts = append(parts, fmt.Sprintf("same session %s", summary.SessionKey))
	}
	if summary.ConversationID != "" {
		parts = append(parts, fmt.Sprintf("conversation %s", summary.ConversationID))
	}
	if summary.PromptCount > 0 {
		parts = append(parts, fmt.Sprintf("%d prompt records", summary.PromptCount))
	}
	if summary.PendingApproval != "" {
		parts = append(parts, fmt.Sprintf("pending approval %s", summary.PendingApproval))
	}
	if summary.ActiveGrantID != "" {
		parts = append(parts, fmt.Sprintf("active grant %s", summary.ActiveGrantID))
	}
	if len(parts) == 0 {
		summary.GroupingRelation = ""
		return
	}
	summary.GroupingRelation = "Grouped by " + strings.Join(parts, ", ") + "."
}

func (r *ChainRepository) loadPromptCoverage(ctx context.Context, summary *api.ChainSummary) error {
	if summary.SessionKey == "" {
		summary.CoveredPrompts = []api.ChainCoveredPrompt{}
		summary.PromptCount = 0
		return nil
	}
	rows, err := r.db.QueryContext(ctx, `
		SELECT
			qa_record_id,
			COALESCE(run_id, ''),
			COALESCE(user_prompt_excerpt, ''),
			COALESCE(risk_level, ''),
			started_at,
			COALESCE(status, '')
		FROM qa_records
		WHERE session_key = ?
		  AND COALESCE(user_prompt_excerpt, '') <> ''
		ORDER BY started_at ASC, qa_record_id ASC
		LIMIT 50`,
		summary.SessionKey,
	)
	if err != nil {
		return err
	}
	defer rows.Close()

	prompts := make([]api.ChainCoveredPrompt, 0)
	for rows.Next() {
		var prompt api.ChainCoveredPrompt
		if err := rows.Scan(
			&prompt.QARecordID,
			&prompt.RunID,
			&prompt.UserPromptExcerpt,
			&prompt.RiskLevel,
			&prompt.StartedAtMs,
			&prompt.Status,
		); err != nil {
			return err
		}
		prompts = append(prompts, prompt)
	}
	if err := rows.Err(); err != nil {
		return err
	}
	summary.CoveredPrompts = prompts
	summary.PromptCount = len(prompts)
	return nil
}

func chainStatus(eventType string) string {
	switch eventType {
	case "agent_end", "session_end", "subagent_ended", "chain_complete":
		return "ended"
	default:
		return "active"
	}
}

func endedAt(eventType string, now string) any {
	if chainStatus(eventType) == "ended" {
		return now
	}
	return nil
}

func truncate(value string, max int) string {
	if len(value) <= max {
		return value
	}
	return value[:max]
}
